pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+// in a separate Go routine. Use `waitForVerifier` to wait for
+// the verifier to finish.
+func (cmvt *cdcMixedVersionTester) setupVerifier(node int) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status(fmt.Sprintf("setting up changefeed verifier for table %s", target))
+
+		// we could just return the error here and let `Wait` return the
+		// error. However, calling t.Fatal directly lets us stop the test
+		// earlier
+		cmvt.monitor.Go(func(ctx context.Context) error {
+			consumer, err := cmvt.kafka.consumer(ctx, "warehouse")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer consumer.Close()
+
+			db := u.conn(ctx, t, node)
+			if _, err := db.Exec(
+				"CREATE TABLE fprint (" +
+					"w_id INT NOT NULL PRIMARY KEY, " +
+					"w_name VARCHAR(10) NOT NULL, " +
+					"w_street_1 VARCHAR(20) NOT NULL, " +
+					"w_street_2 VARCHAR(20) NOT NULL, " +
+					"w_city VARCHAR(20) NOT NULL, " +
+					"w_state VARCHAR(2) NOT NULL, " +
+					"w_zip VARCHAR(9) NOT NULL, " +
+					"w_tax DECIMAL(4, 4) NOT NULL, " +
+					"w_ytd DECIMAL(12, 2) NOT NULL)",
+			); err != nil {
+				t.Fatal(err)
+			}
+
+			fprintV, err := cdctest.NewFingerprintValidator(db, `tpcc.warehouse`, `fprint`, consumer.partitions, 0)
+			if err != nil {
+				t.Fatal(err)
+			}
+			validators := cdctest.Validators{
+				cdctest.NewOrderValidator("tpcc.warehouse"),
+				fprintV,
+			}
+			cmvt.validator = cdctest.MakeCountValidator(validators)
+
+			for {
+				m := consumer.Next(ctx)
+				if m == nil {
+					t.Fatal(fmt.Errorf("unexpected end of changefeed"))
+					return nil
+				}
+
+				updated, resolved, err := cdctest.ParseJSONValueTimestamps(m.Value)
+				if err != nil {
+					t.Fatal(err)
+					return nil
+				}
+
+				partitionStr := strconv.Itoa(int(m.Partition))
+				// errors in the calls to the validator below could lead to
+				// transient failures if the cluster is upgrading, so we retry
+				// for 1 minute.
+				if len(m.Key) > 0 {
+					if err := retry.ForDuration(1*time.Minute, func() error {
+						return cmvt.validator.NoteRow(partitionStr, string(m.Key), string(m.Value), updated)
+					}); err != nil {
+						t.Fatal(err)
+						return nil
+					}
+				} else {
+					if err := retry.ForDuration(1*time.Minute, func() error {
+						return cmvt.validator.NoteResolved(partitionStr, resolved)
+					}); err != nil {
+						t.Fatal(err)
+						return nil
+					}
+					t.L().Printf("%d of %d resolved timestamps validated, latest is %s behind realtime",
+						cmvt.validator.NumResolvedWithRows, requestedResolved, timeutil.Since(resolved.GoTime()))
+
+					if cmvt.validator.NumResolvedWithRows >= requestedResolved {
+						break
+					}
+				}
+			}
+
+			close(cmvt.verifierDone)
+			return nil
+		})
+	}
+}
+
+// validatorVerify checks if the validator has found any issues at the
+// time the function is called.
+func (cmvt *cdcMixedVersionTester) validatorVerify() versionStep {
```

Comment:
Doesn't read like a proper verb or phrase... maybe rename to `validate` or `assertValid`?

Reply:
Renamed to `assertValid`.